chalk-solve/src/infer/unify.rs

```diff
@@ -157,11 +158,11 @@ impl<'t, I: Interner> Unifier<'t, I> {
             | (&TyData::Dyn(_), &TyData::Function(_)) => Err(NoSolution),
 
             (&TyData::Placeholder(ref p1), &TyData::Placeholder(ref p2)) => {
-                Zip::zip_with(self, p1, p2)
+                Zip::zip_with(self, variance, p1, p2)
             }
 
             (&TyData::Apply(ref apply1), &TyData::Apply(ref apply2)) => {
-                Zip::zip_with(self, apply1, apply2)
+                Zip::zip_with(self, variance, apply1, apply2)
```

Comment:
Where things are going to get more interesting is that we need to be applying variance *transformations* at points like this. i suspect the right way to manage this is going to be to remove this impl:

https://github.com/rust-lang/chalk/blob/1781c431c149e18aeee6329c8fd502544efd0f6e/chalk-ir/src/zip.rs#L281-L289

and instead have a specialization function for zipping substitutions that takes some way to "fetch" the variance for each element and which [transforms](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.Variance.html#method.xform) the variance appropriate. Similar to [this code in rustc](https://github.com/rust-lang/rust/blob/7c78a5f97de07a185eebae5a5de436c80d8ba9d4/src/librustc_middle/ty/relate.rs#L136-L150).

Reply:
Yeah...this was the bit that I was unsure about. Still don't know how sure I am. But I can look a bit into how rustc does it.